hs-bindgen/src-internal/HsBindgen/Artefact.hs

```diff
@@ -140,6 +168,21 @@ runArtefacts
       FinalModuleSafe     -> liftIO backendFinalModuleSafe
       FinalModuleUnsafe   -> liftIO backendFinalModuleUnsafe
       FinalModules        -> liftIO backendFinalModules
+      -- File writes.
+      PendingFileWrite what path content -> do
+        fileExists <- liftIO $ doesFileExist path
+        case (fop, content) of
+          (ProtectExistingFiles, _)
+            | fileExists ->
+              throwError (FileAlreadyExistsException path)
+          (_, TextContent c) ->
+            liftIO $ do
+              traceWith tracer $ RunArtefactWriteFile what path
+              createDirectoryIfMissing True $ takeDirectory path
+              writeFile path c
+          (_, BindingSpecContent ubs) ->
+            liftIO $ BindingSpec.writeFile path ubs
```

Comment:
Sorry, that's a nit: Could we maybe move the implementation into a where clause or so? I really like how each branch has only one line, at the moment.